### ISSUE TITLE
[List/Array/Seq].padTo method

### ISSUE THREAD

**Issue by cmorley191** [ntwilson/SafetyFirst#41]
Add a method in the same family as skip or take: padTo extends a sequence to a certain length, using a given constant. e.g.,
```
let padTo count (pad:'a) (lst:'a list) =
  if count <= lst.Length then lst
  else List.append lst (List.replicate (count - lst.Length) pad)
```
A similar method should be supported for NonEmpty sequences.

**Comment by ntwilson** [comment]
Some other alternatives to consider (let's assume `myData = [3; 1; 2]`) :
```
let padTo count (pad : int -> 'a) (lst : 'a list) =
  if count <= List.length lst 
  then lst
  else List.append lst (List.init (count - List.length lst) pad)
...
List.padTo 5 id myData // = [3;1;2;0;1] 
```
```
let padTo count (pad : int -> 'a) (lst : 'a list) = 
  if count <= List.length lst 
  then lst
  else List.append lst (List.init (count - List.length lst) (fun i -> pad (i + Lits.length lst))
...
List.padTo 5 id myData // = [3;1;2;3;4] 
```
```
let zipChoose data defaults = 
  if List.length data >= List.length defaults 
  then data
  else
    let optionalData = Seq.map Some data |> Seq.append (Seq.initInfinite (fun _ -> None))
    Seq.zip optionalData defaults
    |> Seq.map (function | (Some x, _) -> x | (None, dflt) -> dflt)
    |> Seq.toList

...
let defaults = [ for i in 1 .. 5 -> 100 ]
List.zipChoose myData defaults // = [3;1;2;100;100]
```
Not saying those are good implementations (especially for `zipChoose`), but it gets the idea across.  I would also want to think more about the name `zipChoose` if we decided to implement that.

**Comment by ntwilson** [comment]
I'd probably more inclined towards one of the `padTo` implementations that take a lambda, since it's more flexible, and very easy to reproduce the behavior of the one taking in a constant.  I'm not sure though which of the implementations that take a lambda make more sense.  
Not sure what to make of `zipChoose`.  It takes more work to reproduce `padTo`, but is probably the most flexible.  Maybe they're not mutually exclusive...

**Comment by ntwilson** [comment]
I'm also wondering if this belongs in the project LINQED, since it really has nothing to do with safety.  Though I admit the line between the two gets blurred, especially when you want to consider a "safe" version of padTo, which takes a `PositiveInt` as input and returns a NonEmpty sequence.

**Comment by jimboreilly** [comment]
I think the middle implementation is harder to follow. It's still referentially transparent but I'd find it hard to imagine that the 'place' of the data is that valuable over the index/count of values that have been added. This seems _similar_ to [Array.prototype.fill](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/fill) in JS except that it replaces all of the contents, which I find pretty useful if naming is of the question (maybe fillWith? for a lamba)

**Comment by ntwilson** [comment]
Yeah I do want to put some thought into naming, and I like the names fillWith or fillTo or something like that.  Though I'm a little bit concerned if that would seem familiar to a JS programmer and cause confusion since the behavior is actually pretty different (it only adds values, it doesn't replace anything)

**Comment by ntwilson** [comment]
I gave some more thought to this because a related issue came up at work.  I think the most generic function to consider would look like: `outerZip : (Choice<('left * 'right), 'left, 'right> -> 'output>) -> 'left seq -> 'right seq -> 'output seq`.  I'm using the name `outerZip` because compared to the regular zip, it reminds me of an outer join compared to a regular inner join.  I'm not set on that name though.  You could use `outerZip` to directly implement all the others: regular zip, padTo, zipChoose, etc. so I'm inclined to build it because it's the most generic solution.  But I don't think it's a complete solution because it could be pretty cumbersome to use.  For example, to build `padTo`, you'd have to do:
```F#
let padTo length defaultValue xs = 
  outerZip (function 
    | Choice1Of3 (left, right) -> left
    | Choice2Of3 justLeft -> justLeft
    | Choice3Of3 justRight -> justRight)
    xs
    [ for i in 1 .. length -> defaultValue ]
```
Clearly too much work to use outerZip inline - you have to build helper functions out of it.

I'd like to consider also building zipChoose, padTo, and/or an alternate version of padTo which doesn't take a default value, but looks like: `padTo : length -> 'a seq -> 'a option seq`.  (You could easily get the behavior of the other `padTo` function by doing `padTo length |> Seq.map (Option.defaultValue <whatever default you wanted>)`.  I suspect that `None` would be the most common default value anyway.

I'm probably most inclined to build zipChoose (but under a different name), since it's very easy to build `padTo` with it, but unlocks more uses, since you can use different defaults if you need to.

Thoughts are appreciated!
